docs/reference-manual/native-image/DebugInfo.md

```diff
@@ -772,4 +789,158 @@ For example, calling the method below prints high-level information about the Na
+    0.00 :   519d77: nop
+         : 524    java.util.Objects::requireNonNull():
+         : 207    if (obj == null)
+    0.00 :   519d78: nop
+    0.00 :   519d79: nop
+         : 209    return obj;
+    . . .
+```
+
+The leftmost column shows percentages for the amount of time recorded
+at each instruction in samples obtained during the `perf record` run.
+Each instruction is prefaced with it's address in the program's code
+section.
+The disassembly interleaves the source lines from which the code is
+derived, 521-524 for the top level code and 207-209 for the code
+inlined from from `Objects.requireNonNull()`.
+(note, however that the lines for the quoted comments are being computed
+and displayed incorrectly).
+Also, the start of the method is labelled with the name defined in the
+DWARF debug info, `java.lang.String::String()`.
+However, the branch instruction `jbe` at address `0x519d6f` uses a
+very large offset from `graal_vm_locator_symbol` to identify what is
+actually a target address within the compiled code for
+`String::String()`.
+
+Readability of the tool output is significantly improved if
+option `-H-DeleteLocalSymbols` is passed to the `native-image`
+command.
+The equivalent `perf annotate` output with this option enabled is as
+follows:
+
+```
+    . . .
+         : 5      000000000051aac0 <String_constructor_f60263d569497f1facccd5467ef60532e990f75d>:
+         : 6      java.lang.String::String():
+         : 521    *          {@code offset} is greater than {@code bytes.length - length}
+         : 522    *
+         : 523    * @since  1.6
+         : 524    */
+         : 525    @SuppressWarnings("removal")
+         : 526    public String(byte[] bytes, int offset, int length, Charset charset) {
+    0.00 :   51aac0: sub    $0x68,%rsp
+    0.00 :   51aac4: mov    %rdi,0x38(%rsp)
+    0.00 :   51aac9: mov    %rsi,0x30(%rsp)
+    0.00 :   51aace: mov    %edx,0x64(%rsp)
+    0.00 :   51aad2: mov    %ecx,0x60(%rsp)
+    0.00 :   51aad6: mov    %r8,0x28(%rsp)
+    0.00 :   51aadb: cmp    0x8(%r15),%rsp
+    0.00 :   51aadf: jbe    51bbc1 <String_constructor_f60263d569497f1facccd5467ef60532e990f75d+0x1101>
+    0.00 :   51aae5: nop
+    0.00 :   51aae6: nop
+         : 522    Objects.requireNonNull(charset);
+    0.00 :   51aae7: nop
+         : 524    java.util.Objects::requireNonNull():
+         : 207    * @param <T> the type of the reference
+         : 208    * @return {@code obj} if not {@code null}
+         : 209    * @throws NullPointerException if {@code obj} is {@code null}
+         : 210    */
+         : 211    public static <T> T requireNonNull(T obj) {
+         : 212    if (obj == null)
+    0.00 :   51aae8: nop
+    0.00 :   51aae9: nop
+         : 209    throw new NullPointerException();
+         : 210    return obj;
+    . . .
+```
+
+Note that the start address of the method is now labelled with the
+mangled symbol name `String_constructor_f60263d569497f1facccd5467ef60532e990f75d` as well as the DWARF name.
```

Comment:
Would be cool if `c++filt` would support our mangling.

Reply:
Yes it would. I have had a quick lopk into what perf does to unmangle names. By default it calls `bfd_demangle` which has its own slightly weird expectations about how to handle names e.g. arrays of `TYPE` are modelled as `JArray<TYPE>`. If (when) that fails it should punt to a routine in the perf code called `java_demangle_sym`. This knows how to demangle bytecode type and member signatures into Java syntax. For example it will turn `Ljava/lang/StringLatin1;equals([B[B)Z` into `boolean java.lang.StringLatin1.equals(byte[], byte[])"`. We might be able to use that for some cases but I guess it won't do for methods exposed as C entry points which actually need a linker-friendly name.

Note that anyway I am not currently using the symbol name as the `linkage_name` atttribute in the method.definition. At present I generate that attribute as  `<fully_qualified_class_name>::<bare_method_name>` in order to ensure that 1) `perf report` prints something sensible when there are no local symbols 2) `perf annotate` includes this name at the start of the listing for a given method. We can probably do better than this (but maybe in a follow-up patch).

Reply:
> but maybe in a follow-up patch

Definitely